File: src/main/java/com/wafflestudio/draft/service/ParticipantService.java
package com.wafflestudio.draft.service;

import com.wafflestudio.draft.dto.response.ParticipantsResponse;
import com.wafflestudio.draft.dto.response.UserInformationResponse;
import com.wafflestudio.draft.model.Participant;
import com.wafflestudio.draft.model.Room;
import com.wafflestudio.draft.model.User;
import com.wafflestudio.draft.repository.ParticipantRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
@RequiredArgsConstructor
public class ParticipantService {
    private final ParticipantRepository participantRepository;

    public ParticipantsResponse getParticipants(Room room) {
        List<UserInformationResponse> participantsOfTeam1 = participantRepository.getUsersInTeam(room, 1);
        List<UserInformationResponse> participantsOfTeam2 = participantRepository.getUsersInTeam(room, 2);
        return new ParticipantsResponse(participantsOfTeam1, participantsOfTeam2);
    }

    public ParticipantsResponse addParticipants(Room room, User user) {
        List<UserInformationResponse> participantsOfTeam1 = participantRepository.getUsersInTeam(room, 1);
        List<UserInformationResponse> participantsOfTeam2 = participantRepository.getUsersInTeam(room, 2);
        int teamOfNewParticipant = 1;
        if (participantsOfTeam1.size() > participantsOfTeam2.size()) {
            teamOfNewParticipant = 2;
            participantsOfTeam2.add(new UserInformationResponse(user));
        } else {
            participantsOfTeam1.add(new UserInformationResponse(user));
        }
        Participant newParticipant = new Participant(user, room, teamOfNewParticipant);
        participantRepository.save(newParticipant);
        return new ParticipantsResponse(participantsOfTeam1, participantsOfTeam2);
    }
}

File: src/main/kotlin/com/wafflestudio/draft/service/ParticipantService.kt
package com.wafflestudio.draft.service

import com.wafflestudio.draft.dto.response.ParticipantsResponse
import com.wafflestudio.draft.dto.response.UserInformationResponse
import com.wafflestudio.draft.model.Participant
import com.wafflestudio.draft.model.Room
import com.wafflestudio.draft.model.User
import com.wafflestudio.draft.model.enums.Team
import com.wafflestudio.draft.repository.ParticipantRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class ParticipantService(private val participantRepository: ParticipantRepository) {
    fun getParticipants(room: Room?): ParticipantsResponse {
        val participantsOfTeam1: MutableList<UserInformationResponse>? = participantRepository.getUsersInTeam(room, Team.A)
        val participantsOfTeam2: MutableList<UserInformationResponse>? = participantRepository.getUsersInTeam(room, Team.B)
        return ParticipantsResponse(participantsOfTeam1, participantsOfTeam2)
    }

    fun addParticipants(room: Room, user: User): ParticipantsResponse {
        val participantsOfTeam1 = participantRepository.getUsersInTeam(room, Team.A)
        val participantsOfTeam2 = participantRepository.getUsersInTeam(room, Team.B)
        var teamOfNewParticipant = Team.A
        if (participantsOfTeam1!!.size > participantsOfTeam2!!.size) {
            teamOfNewParticipant = Team.B
            participantsOfTeam2.add(UserInformationResponse(user))
        } else {
            participantsOfTeam1.add(UserInformationResponse(user))
        }
        val newParticipant = Participant(user, room, teamOfNewParticipant)
        participantRepository.save(newParticipant)
        return ParticipantsResponse(participantsOfTeam1, participantsOfTeam2)
    }

    fun deleteParticipants(room: Room, user: User) {
        participantRepository.deleteParticipantByRoomAndUser(room, user)
    }
}
